### PR TITLE
LibJS: Implement standard semantics for relational operators

### PR DESCRIPTION
**Pull request by nooga**
Previously, the relational operators where casting any value to double and comparing the results according to C++ semantics.

This patch makes the relational operators in JS behave according to the standard specification.

Since we don't have BigInt yet, the implementation doesn't take it into account.

### CODE REVIEW

File: Libraries/LibJS/Runtime/Value.cpp [hunk 1]
```diff
@@ -673,4 +662,77 @@ bool abstract_eq(Interpreter& interpreter, Value lhs, Value rhs)
     return false;
 }
 
+TriState abstract_relation(Interpreter& interpreter, bool left_first, Value lhs, Value rhs)
+{
+    Value x_primitive = {};
+    Value y_primitive = {};
+
+    if (left_first) {
+        x_primitive = lhs.to_primitive(interpreter);
```
Comment (linusg) [inline Libraries/LibJS/Runtime/Value.cpp:671]: I think we should update `Value::to_primitive()` to take a `PreferredType` and pass that on to `object::to_primitive()`and then be explicit and give it the number hint mentioned in the spec.

File: Libraries/LibJS/Runtime/Value.cpp [hunk 2]
```diff
@@ -673,4 +662,77 @@ bool abstract_eq(Interpreter& interpreter, Value lhs, Value rhs)
...
+    } else {
+        y_primitive = lhs.to_primitive(interpreter);
+        if (interpreter.exception())
+            return {};
+        x_primitive = rhs.to_primitive(interpreter);
+        if (interpreter.exception())
+            return {};
+    }
+
+    if (x_primitive.type() == Value::Type::String && y_primitive.type() == Value::Type::String) {
```
Comment (linusg) [inline Libraries/LibJS/Runtime/Value.cpp:686]:
```suggestion
    if (x_primitive.is_string() && y_primitive.is_string()) {
```

File: Libraries/LibJS/Runtime/Value.cpp [hunk 3]
```diff
@@ -673,4 +662,77 @@ bool abstract_eq(Interpreter& interpreter, Value lhs, Value rhs)
...
+            return TriState::False;
+        if (y_string.starts_with(x_string))
+            return TriState::True;
+
+        Utf8View x_codepoints { x_string };
+        Utf8View y_codepoints { y_string };
+        for (auto k = x_codepoints.begin(), l = y_codepoints.begin();
+             k != x_codepoints.end() && l != y_codepoints.end();
+             ++k, ++l) {
+            if (*k != *l) {
```
Comment (linusg) [inline Libraries/LibJS/Runtime/Value.cpp:700]: this looks weird - is clang-format doing this?

File: Libraries/LibJS/Runtime/Value.cpp [hunk 4]
```diff
@@ -673,4 +662,77 @@ bool abstract_eq(Interpreter& interpreter, Value lhs, Value rhs)
...
+                    return TriState::False;
+                }
+            }
+        }
+        ASSERT_NOT_REACHED();
+    }
+
+    // FIXME add BigInt cases here once we have BigInt
+
+    Value x_numeric = x_primitive.to_number(interpreter);
```
Comment (linusg) [inline Libraries/LibJS/Runtime/Value.cpp:713]: Not that it matters that much, but I'd say Serenity code should try to leverage `auto` as much as possible - 

```suggestion
    auto x_numeric = x_primitive.to_number(interpreter);
```

same for `double x_value = x_numeric.as_double();` - looks like Java :smile:

File: Libraries/LibJS/Runtime/ArrayPrototype.cpp [hunk 5]
```diff
@@ -706,8 +706,7 @@ Value ArrayPrototype::splice(Interpreter& interpreter)
                 this_object->put(to, from);
                 if (interpreter.exception())
                     return {};
-            }
-            else
+            } else
```
[the remark this answers is not in the document]
Comment (nooga) [reply Libraries/LibJS/Runtime/ArrayPrototype.cpp:709]: I rebased onto the master and ran clang-format so this just happened :D

File: Libraries/LibJS/Runtime/Value.cpp (hunk 1 above)
Comment (nooga) [reply Libraries/LibJS/Runtime/Value.cpp:671]: Yeah, that would be useful. Would you like to do it as a part of this PR?

File: Libraries/LibJS/Runtime/Value.cpp (hunk 3 above)
Comment (nooga) [reply Libraries/LibJS/Runtime/Value.cpp:700]: I was unsure if we want a super long line so I've broken it up, clang-format left it as it is.

Comment (linusg) [reply Libraries/LibJS/Runtime/Value.cpp:700]: either way is fine :+1:

File: Libraries/LibJS/Runtime/Value.cpp (hunk 1 above)
Comment (linusg) [reply Libraries/LibJS/Runtime/Value.cpp:671]: sure, if you don't mind :)

**Comment by nooga** [comment]
@linusg please have a look. I've moved some things around in order to handle the preferred type.

**Comment by nooga** [comment]
Going to squash them anyway.

**Pull request merged**